WayName is shown even when the text is empty after recentering

<!--
Hello and thanks for contributing! To help us diagnose your problem quickly, please:

 - Include a minimal demonstration of the bug, including code, logs, and screenshots.
 - Ensure you can reproduce the bug using the latest release.
 - Only post to report a bug or request a feature; direct all other questions to: https://stackoverflow.com/questions/tagged/mapbox
-->

**Android API:** 28
**Mapbox Navigation SDK version:** 0.41.0

The `WayName` view can become visible even when there's no text, resulting in a white bubble being shown.

Normally, this check is made in `NavigationPresenter`:
https://github.com/mapbox/mapbox-navigation-android/blob/f2a082cfd33792f2209c67979f615668d3f7a587/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationPresenter.java#L71-L75

However, when recentering, the check is not made:
https://github.com/mapbox/mapbox-navigation-android/blob/f2a082cfd33792f2209c67979f615668d3f7a587/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationPresenter.java#L24-L27

Thus, recentering causes the WayName view to always be shown.

@Guardiola31337 I can make a PR if you want, but I'd just add the same check as in the first code example. Maybe there's a better way to do it, so you don't have to remember to add the check again if the WayName visibility is changed somewhere else.

I can't start the test app, but the reproduction steps is to start the navigation on a road with no name, move the map, and press the recenter button. In my case, our work place is situated between a bunch of small roads (of a parking lot) with no names, so every time I test navigation, it happens.

For some reason I'm not able to start the test app. I get the following error:

```
    --------- beginning of crash
2019-08-26 09:10:53.376 31797-31797/com.mapbox.services.android.navigation.app E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.mapbox.services.android.navigation.app, PID: 31797
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.mapbox.services.android.navigation.app/com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity}: java.lang.RuntimeException: Cannot create an instance of class com.mapbox.services.android.navigation.testapp.example.ui.ExampleViewModel
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3047)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3182)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1916)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6898)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.RuntimeException: Cannot create an instance of class com.mapbox.services.android.navigation.testapp.example.ui.ExampleViewModel
        at android.arch.lifecycle.ViewModelProvider$AndroidViewModelFactory.create(ViewModelProvider.java:207)
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:134)
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:102)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$viewModel$2.invoke(ExampleActivity.kt:46)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$viewModel$2.invoke(ExampleActivity.kt:43)
        at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.getViewModel(Unknown Source:7)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.access$getViewModel$p(ExampleActivity.kt:43)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$presenter$2.invoke(ExampleActivity.kt:49)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$presenter$2.invoke(ExampleActivity.kt:43)
        at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.getPresenter(Unknown Source:7)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.setupWith(ExampleActivity.kt:292)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.onCreate(ExampleActivity.kt:55)
        at android.app.Activity.performCreate(Activity.java:7149)
        at android.app.Activity.performCreate(Activity.java:7140)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1288)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3027)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3182) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1916) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6898) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.arch.lifecycle.ViewModelProvider$AndroidViewModelFactory.create(ViewModelProvider.java:199)
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:134) 
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:102) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$viewModel$2.invoke(ExampleActivity.kt:46) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$viewModel$2.invoke(ExampleActivity.kt:43) 
        at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.getViewModel(Unknown Source:7) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.access$getViewModel$p(ExampleActivity.kt:43) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$presenter$2.invoke(ExampleActivity.kt:49) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$presenter$2.invoke(ExampleActivity.kt:43) 
        at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.getPresenter(Unknown Source:7) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.setupWith(ExampleActivity.kt:292) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.onCreate(ExampleActivity.kt:55) 
        at android.app.Activity.performCreate(Activity.java:7149) 
        at android.app.Activity.performCreate(Activity.java:7140) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1288) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3027) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3182) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1916) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6898) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
     Caused by: com.mapbox.services.android.navigation.v5.exception.NavigationException: A valid access token must be passed in when first initializing MapboxNavigation
        at com.mapbox.services.android.navigation.v5.navigation.NavigationTelemetry.validateAccessToken(NavigationTelemetry.java:304)
        at com.mapbox.services.android.navigation.v5.navigation.NavigationTelemetry.initialize(NavigationTelemetry.java:120)
        at com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.initializeTelemetry(MapboxNavigation.java:906)
2019-08-26 09:10:53.376 31797-31797/com.mapbox.services.android.navigation.app E/AndroidRuntime:     at com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.initialize(MapboxNavigation.java:887)
        at com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.<init>(MapboxNavigation.java:118)
        at com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.<init>(MapboxNavigation.java:92)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleViewModel.<init>(ExampleViewModel.kt:64)
        	... 31 more
2019-08-26 09:10:53.377 31797-31797/com.mapbox.services.android.navigation.app E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.mapbox.services.android.navigation.app, PID: 31797
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.mapbox.services.android.navigation.app/com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity}: java.lang.RuntimeException: Cannot create an instance of class com.mapbox.services.android.navigation.testapp.example.ui.ExampleViewModel
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3047)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3182)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1916)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6898)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.RuntimeException: Cannot create an instance of class com.mapbox.services.android.navigation.testapp.example.ui.ExampleViewModel
        at android.arch.lifecycle.ViewModelProvider$AndroidViewModelFactory.create(ViewModelProvider.java:207)
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:134)
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:102)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$viewModel$2.invoke(ExampleActivity.kt:46)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$viewModel$2.invoke(ExampleActivity.kt:43)
        at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.getViewModel(Unknown Source:7)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.access$getViewModel$p(ExampleActivity.kt:43)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$presenter$2.invoke(ExampleActivity.kt:49)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$presenter$2.invoke(ExampleActivity.kt:43)
        at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.getPresenter(Unknown Source:7)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.setupWith(ExampleActivity.kt:292)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.onCreate(ExampleActivity.kt:55)
        at android.app.Activity.performCreate(Activity.java:7149)
        at android.app.Activity.performCreate(Activity.java:7140)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1288)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3027)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3182) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1916) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6898) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.arch.lifecycle.ViewModelProvider$AndroidViewModelFactory.create(ViewModelProvider.java:199)
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:134) 
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:102) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$viewModel$2.invoke(ExampleActivity.kt:46) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$viewModel$2.invoke(ExampleActivity.kt:43) 
        at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.getViewModel(Unknown Source:7) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.access$getViewModel$p(ExampleActivity.kt:43) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$presenter$2.invoke(ExampleActivity.kt:49) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity$presenter$2.invoke(ExampleActivity.kt:43) 
        at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.getPresenter(Unknown Source:7) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.setupWith(ExampleActivity.kt:292) 
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleActivity.onCreate(ExampleActivity.kt:55) 
        at android.app.Activity.performCreate(Activity.java:7149) 
        at android.app.Activity.performCreate(Activity.java:7140) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1288) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3027) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3182) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1916) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6898) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
     Caused by: com.mapbox.services.android.navigation.v5.exception.NavigationException: A valid access token must be passed in when first initializing MapboxNavigation
        at com.mapbox.services.android.navigation.v5.navigation.NavigationTelemetry.validateAccessToken(NavigationTelemetry.java:304)
        at com.mapbox.services.android.navigation.v5.navigation.NavigationTelemetry.initialize(NavigationTelemetry.java:120)
        at com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.initializeTelemetry(MapboxNavigation.java:906)
2019-08-26 09:10:53.377 31797-31797/com.mapbox.services.android.navigation.app E/AndroidRuntime:     at com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.initialize(MapboxNavigation.java:887)
        at com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.<init>(MapboxNavigation.java:118)
        at com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.<init>(MapboxNavigation.java:92)
        at com.mapbox.services.android.navigation.testapp.example.ui.ExampleViewModel.<init>(ExampleViewModel.kt:64)
```

--- a/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationContract.java
+++ b/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationContract.java
@@ -17,6 +17,13 @@ public interface NavigationContract {
 
     boolean isSummaryBottomSheetHidden();
 
+    void setWayNameActive(boolean isActive);
+
+    void setWayNameVisibility(boolean isVisible);
+
+    String retrieveWayNameText();
+
+    @Deprecated
     void updateWayNameVisibility(boolean isVisible);
 
     void updateWayNameView(@NonNull String wayName);

--- a/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationPresenter.java
+++ b/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationPresenter.java
@@ -25,7 +25,10 @@ class NavigationPresenter {
   void onRecenterClick() {
     view.setSummaryBehaviorHideable(false);
     view.setSummaryBehaviorState(BottomSheetBehavior.STATE_EXPANDED);
-    view.updateWayNameVisibility(true);
+    view.setWayNameActive(true);
+    if (!TextUtils.isEmpty(view.retrieveWayNameText())) {
+      view.setWayNameVisibility(true);
+    }
     view.resetCameraPosition();
     view.hideRecenterBtn();
   }
@@ -34,7 +37,8 @@ class NavigationPresenter {
     if (!view.isSummaryBottomSheetHidden()) {
       view.setSummaryBehaviorHideable(true);
       view.setSummaryBehaviorState(BottomSheetBehavior.STATE_HIDDEN);
-      view.updateWayNameVisibility(false);
+      view.setWayNameActive(false);
+      view.setWayNameVisibility(false);
     }
   }
 
@@ -71,19 +75,23 @@ class NavigationPresenter {
 
   void onWayNameChanged(@NonNull String wayName) {
     if (TextUtils.isEmpty(wayName) || view.isSummaryBottomSheetHidden()) {
-      view.updateWayNameVisibility(false);
+      view.setWayNameActive(false);
+      view.setWayNameVisibility(false);
       return;
     }
     view.updateWayNameView(wayName);
-    view.updateWayNameVisibility(true);
+    view.setWayNameActive(true);
+    view.setWayNameVisibility(true);
   }
 
   void onNavigationStopped() {
-    view.updateWayNameVisibility(false);
+    view.setWayNameActive(false);
+    view.setWayNameVisibility(false);
   }
 
   void onRouteOverviewClick() {
-    view.updateWayNameVisibility(false);
+    view.setWayNameActive(false);
+    view.setWayNameVisibility(false);
     view.updateCameraRouteOverview();
     view.showRecenterBtn();
   }

--- a/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java
+++ b/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java
@@ -320,6 +320,39 @@ public class NavigationView extends CoordinatorLayout implements LifecycleOwner,
   }
 
   /**
+   * Set the auto-query state to provide the default way names value
+   * with {@link NavigationMapboxMap#updateWaynameQueryMap(boolean)}.
+   *
+   * @param isActive true if auto-query is enabled, false otherwise.
+   */
+  @Override
+  public void setWayNameActive(boolean isActive) {
+    if (navigationMap != null) {
+      navigationMap.updateWaynameQueryMap(isActive);
+    }
+  }
+
+  /**
+   * Set the visibility state of the way name view.
+   *
+   * @param isVisible true if visible, false otherwise.
+   */
+  @Override
+  public void setWayNameVisibility(boolean isVisible) {
+    wayNameView.updateVisibility(isVisible);
+  }
+
+  /**
+   * Provides the current text of the way name view.
+   *
+   * @return the current text of the way name view
+   */
+  @Override
+  public String retrieveWayNameText() {
+    return wayNameView.retrieveWayNameText();
+  }
+
+  /**
    * Updates the visibility of the way name view that is show below
    * the navigation icon.
    * <p>
@@ -328,8 +361,16 @@ public class NavigationView extends CoordinatorLayout implements LifecycleOwner,
    * {@link NavigationMapboxMap#updateWaynameQueryMap(boolean)}.
    *
    * @param isVisible true to show, false to hide
+   *
+   * @deprecated This method sets the auto-query, by calling
+   * {@link NavigationMapboxMap#updateWaynameQueryMap(boolean)}, and updates
+   * the visibility of the way name view at the same time. We recommend to use {@link #setWayNameActive(boolean)}
+   * to enable auto-query or {@link #setWayNameVisibility(boolean)} to update the
+   * way name view visibility. Remember to check whether the way name view text is empty or not before
+   * set the auto-query. {@link #retrieveWayNameText()} could get the way name view text.
    */
   @Override
+  @Deprecated
   public void updateWayNameVisibility(boolean isVisible) {
     wayNameView.updateVisibility(isVisible);
     if (navigationMap != null) {

--- a/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/OnWayNameChangedListener.java
+++ b/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/OnWayNameChangedListener.java
@@ -8,7 +8,7 @@ import androidx.annotation.NonNull;
  * <p>
  * This listener is triggered when a new way name is found along the route.  It will be triggered
  * regardless of the map way name visibility
- * ({@link com.mapbox.services.android.navigation.ui.v5.NavigationView#updateWayNameVisibility(boolean)}).
+ * ({@link com.mapbox.services.android.navigation.ui.v5.NavigationView#setWayNameActive(boolean)}).
  * This is so you can hide our implementation of the way name UI and update your own if you'd like.
  */
 public interface OnWayNameChangedListener {

--- a/libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/NavigationPresenterTest.java
+++ b/libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/NavigationPresenterTest.java
@@ -37,7 +37,8 @@ public class NavigationPresenterTest {
 
     presenter.onRouteOverviewClick();
 
-    verify(view).updateWayNameVisibility(false);
+    verify(view).setWayNameActive(false);
+    verify(view).setWayNameVisibility(false);
   }
 
   @Test
@@ -63,11 +64,13 @@ public class NavigationPresenterTest {
   @Test
   public void onRecenterBtnClick_mapWayNameIsShown() {
     NavigationContract.View view = mock(NavigationContract.View.class);
+    when(view.retrieveWayNameText()).thenReturn("Some way name");
     NavigationPresenter presenter = new NavigationPresenter(view);
 
     presenter.onRecenterClick();
 
-    verify(view).updateWayNameVisibility(true);
+    verify(view).setWayNameActive(true);
+    verify(view).setWayNameVisibility(true);
   }
 
   @Test
@@ -77,7 +80,8 @@ public class NavigationPresenterTest {
 
     presenter.onWayNameChanged("Some way name");
 
-    verify(view).updateWayNameVisibility(true);
+    verify(view).setWayNameActive(true);
+    verify(view).setWayNameVisibility(true);
   }
 
   @Test
@@ -98,7 +102,8 @@ public class NavigationPresenterTest {
 
     presenter.onWayNameChanged("");
 
-    verify(view).updateWayNameVisibility(false);
+    verify(view).setWayNameActive(false);
+    verify(view).setWayNameVisibility(false);
   }
 
   @Test
@@ -109,7 +114,8 @@ public class NavigationPresenterTest {
 
     presenter.onWayNameChanged("some valid way name");
 
-    verify(view).updateWayNameVisibility(false);
+    verify(view).setWayNameActive(false);
+    verify(view).setWayNameVisibility(false);
   }
 
   @Test
@@ -119,7 +125,8 @@ public class NavigationPresenterTest {
 
     presenter.onNavigationStopped();
 
-    verify(view).updateWayNameVisibility(false);
+    verify(view).setWayNameActive(false);
+    verify(view).setWayNameVisibility(false);
   }
 
   @Test

--- a/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationContract.java
+++ b/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationContract.java
@@ -18,6 +18,13 @@ public interface NavigationContract {
 
     boolean isSummaryBottomSheetHidden();
 
+    void setWayNameActive(boolean isActive);
+
+    void setWayNameVisibility(boolean isVisible);
+
+    String retrieveWayNameText();
+
+    @Deprecated
     void updateWayNameVisibility(boolean isVisible);
 
     void updateWayNameView(@NonNull String wayName);

--- a/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationPresenter.java
+++ b/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationPresenter.java
@@ -26,7 +26,10 @@ class NavigationPresenter {
   void onRecenterClick() {
     view.setSummaryBehaviorHideable(false);
     view.setSummaryBehaviorState(BottomSheetBehavior.STATE_EXPANDED);
-    view.updateWayNameVisibility(true);
+    view.setWayNameActive(true);
+    if (!TextUtils.isEmpty(view.retrieveWayNameText())) {
+      view.setWayNameVisibility(true);
+    }
     view.resetCameraPosition();
     view.hideRecenterBtn();
   }
@@ -35,7 +38,8 @@ class NavigationPresenter {
     if (!view.isSummaryBottomSheetHidden()) {
       view.setSummaryBehaviorHideable(true);
       view.setSummaryBehaviorState(BottomSheetBehavior.STATE_HIDDEN);
-      view.updateWayNameVisibility(false);
+      view.setWayNameActive(false);
+      view.setWayNameVisibility(false);
     }
   }
 
@@ -72,19 +76,23 @@ class NavigationPresenter {
 
   void onWayNameChanged(@NonNull String wayName) {
     if (TextUtils.isEmpty(wayName) || view.isSummaryBottomSheetHidden()) {
-      view.updateWayNameVisibility(false);
+      view.setWayNameActive(false);
+      view.setWayNameVisibility(false);
       return;
     }
     view.updateWayNameView(wayName);
-    view.updateWayNameVisibility(true);
+    view.setWayNameActive(true);
+    view.setWayNameVisibility(true);
   }
 
   void onNavigationStopped() {
-    view.updateWayNameVisibility(false);
+    view.setWayNameActive(false);
+    view.setWayNameVisibility(false);
   }
 
   void onRouteOverviewClick() {
-    view.updateWayNameVisibility(false);
+    view.setWayNameActive(false);
+    view.setWayNameVisibility(false);
     view.updateCameraRouteOverview();
     view.showRecenterBtn();
   }

--- a/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationView.java
+++ b/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationView.java
@@ -325,6 +325,38 @@ public class NavigationView extends CoordinatorLayout implements LifecycleOwner,
   }
 
   /**
+   * Set the auto-query state to provide the default way names value
+   * with {@link NavigationMapboxMap#updateWaynameQueryMap(boolean)}.
+   * @param isActive true if auto-query is enabled, false otherwise.
+   */
+  @Override
+  public void setWayNameActive(boolean isActive) {
+    if (navigationMap != null) {
+      navigationMap.updateWaynameQueryMap(isActive);
+    }
+  }
+
+  /**
+   * Set the visibility state of the way name view.
+   *
+   * @param isVisible true if visible, false otherwise.
+   */
+  @Override
+  public void setWayNameVisibility(boolean isVisible) {
+    wayNameView.updateVisibility(isVisible);
+  }
+
+  /**
+   * Provides the current text of the way name view.
+   *
+   * @return the current text of the way name view
+   */
+  @Override
+  public String retrieveWayNameText() {
+    return wayNameView.retrieveWayNameText();
+  }
+
+  /**
    * Updates the visibility of the way name view that is show below
    * the navigation icon.
    * <p>
@@ -333,8 +365,16 @@ public class NavigationView extends CoordinatorLayout implements LifecycleOwner,
    * {@link NavigationMapboxMap#updateWaynameQueryMap(boolean)}.
    *
    * @param isVisible true to show, false to hide
+   *
+   * @deprecated This method sets the auto-query, by calling
+   * {@link NavigationMapboxMap#updateWaynameQueryMap(boolean)}, and updates
+   * the visibility of the way name view at the same time. We recommend to use {@link #setWayNameActive(boolean)}
+   * to enable auto-query or {@link #setWayNameVisibility(boolean)} to update the
+   * way name view visibility. Remember to check whether the way name view text is empty or not before
+   * set the auto-query. {@link #retrieveWayNameText()} could get the way name view text.
    */
   @Override
+  @Deprecated
   public void updateWayNameVisibility(boolean isVisible) {
     wayNameView.updateVisibility(isVisible);
     if (navigationMap != null) {

--- a/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/map/MapWayName.java
+++ b/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/map/MapWayName.java
@@ -21,7 +21,7 @@ class MapWayName {
 
   private static final String NAME_PROPERTY = "name";
   private static final String EMPTY_CURRENT_WAY_NAME = "";
-  private final MapWaynameProgressChangeListener progressChangeListener = new MapWaynameProgressChangeListener(this);
+  private final MapWayNameChangeObserver mapWayNameChangeObserver = new MapWayNameChangeObserver(this);
   private final Set<OnWayNameChangedListener> onWayNameChangedListeners;
   private WaynameFeatureFinder featureInteractor;
   private List<Point> currentStepPoints = new ArrayList<>();
@@ -91,15 +91,15 @@ class MapWayName {
 
   private void registerObservers() {
     if (navigation != null) {
-      navigation.registerRouteProgressObserver(progressChangeListener);
-      navigation.registerLocationObserver(progressChangeListener);
+      navigation.registerRouteProgressObserver(mapWayNameChangeObserver);
+      navigation.registerLocationObserver(mapWayNameChangeObserver);
     }
   }
 
   private void unregisterObservers() {
     if (navigation != null) {
-      navigation.unregisterRouteProgressObserver(progressChangeListener);
-      navigation.unregisterLocationObserver(progressChangeListener);
+      navigation.unregisterRouteProgressObserver(mapWayNameChangeObserver);
+      navigation.unregisterLocationObserver(mapWayNameChangeObserver);
     }
   }
 

--- a/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/map/MapWayNameChangeObserver.java
+++ b/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/map/MapWayNameChangeObserver.java
@@ -10,11 +10,11 @@ import org.jetbrains.annotations.NotNull;
 
 import java.util.List;
 
-class MapWaynameProgressChangeListener implements RouteProgressObserver, LocationObserver {
+class MapWayNameChangeObserver implements RouteProgressObserver, LocationObserver {
 
   private final MapWayName mapWayName;
 
-  MapWaynameProgressChangeListener(MapWayName mapWayName) {
+  MapWayNameChangeObserver(MapWayName mapWayName) {
     this.mapWayName = mapWayName;
   }
 

--- a/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/map/OnWayNameChangedListener.java
+++ b/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/map/OnWayNameChangedListener.java
@@ -10,7 +10,7 @@ import com.mapbox.navigation.ui.NavigationView;
  * <p>
  * This listener is triggered when a new way name is found along the route.  It will be triggered
  * regardless of the map way name visibility
- * ({@link NavigationView#updateWayNameVisibility(boolean)}).
+ * ({@link NavigationView#setWayNameActive(boolean)}).
  * This is so you can hide our implementation of the way name UI and update your own if you'd like.
  */
 public interface OnWayNameChangedListener {

--- a/libnavigation-ui/src/test/java/com/mapbox/navigation/ui/NavigationPresenterTest.java
+++ b/libnavigation-ui/src/test/java/com/mapbox/navigation/ui/NavigationPresenterTest.java
@@ -1,8 +1,6 @@
 package com.mapbox.navigation.ui;
 
 import com.mapbox.api.directions.v5.models.DirectionsRoute;
-import com.mapbox.navigation.ui.NavigationContract;
-import com.mapbox.navigation.ui.NavigationPresenter;
 
 import org.junit.Test;
 
@@ -39,7 +37,8 @@ public class NavigationPresenterTest {
 
     presenter.onRouteOverviewClick();
 
-    verify(view).updateWayNameVisibility(false);
+    verify(view).setWayNameActive(false);
+    verify(view).setWayNameVisibility(false);
   }
 
   @Test
@@ -65,11 +64,13 @@ public class NavigationPresenterTest {
   @Test
   public void onRecenterBtnClick_mapWayNameIsShown() {
     NavigationContract.View view = mock(NavigationContract.View.class);
+    when(view.retrieveWayNameText()).thenReturn("Some way name");
     NavigationPresenter presenter = new NavigationPresenter(view);
 
     presenter.onRecenterClick();
 
-    verify(view).updateWayNameVisibility(true);
+    verify(view).setWayNameActive(true);
+    verify(view).setWayNameVisibility(true);
   }
 
   @Test
@@ -79,7 +80,8 @@ public class NavigationPresenterTest {
 
     presenter.onWayNameChanged("Some way name");
 
-    verify(view).updateWayNameVisibility(true);
+    verify(view).setWayNameActive(true);
+    verify(view).setWayNameVisibility(true);
   }
 
   @Test
@@ -100,7 +102,8 @@ public class NavigationPresenterTest {
 
     presenter.onWayNameChanged("");
 
-    verify(view).updateWayNameVisibility(false);
+    verify(view).setWayNameActive(false);
+    verify(view).setWayNameVisibility(false);
   }
 
   @Test
@@ -111,7 +114,8 @@ public class NavigationPresenterTest {
 
     presenter.onWayNameChanged("some valid way name");
 
-    verify(view).updateWayNameVisibility(false);
+    verify(view).setWayNameActive(false);
+    verify(view).setWayNameVisibility(false);
   }
 
   @Test
@@ -121,7 +125,8 @@ public class NavigationPresenterTest {
 
     presenter.onNavigationStopped();
 
-    verify(view).updateWayNameVisibility(false);
+    verify(view).setWayNameActive(false);
+    verify(view).setWayNameVisibility(false);
   }
 
   @Test

